bokehjs/src/lib/embed/server.ts

```diff
@@ -8,9 +8,10 @@ import type {EmbedTarget} from "./dom"
 
 // @internal
 export function _get_ws_url(app_path: string | undefined, absolute_url: string | undefined): string {
-  let protocol = "ws:"
-  if (window.location.protocol == "https:") {
-    protocol = "wss:"
+  // if in an `srcdoc` iframe, try to get the absolute URL
+  // from the `data-absolute-url` attribute if not passed explicitly
+  if (absolute_url === undefined && frameElement !== null && (frameElement as HTMLIFrameElement).dataset.absoluteUrl !== undefined) {
+    absolute_url = (frameElement as HTMLIFrameElement).dataset.absoluteUrl
```

Comment:
Let's remove all this casting with:
```ts
  if (absolute_url === undefined && frameElement instanceof HTMLElement && frameElement.dataset.absoluteUrl !== undefined) {
    absolute_url = frameElement.dataset.absoluteUrl
  }
```
or alternatively use `HTMLIFrameElement` instead of `HTMLElement`, though I'm not sure if we need to be that specific.

Reply:
My thinking was the same here! But it does not work. This is because `frameElement` is an `HTMLElement`, but not `window.HTMLElement`. It is an `HTMLElement` from the context of the parent frame. I can make this work without type cast by checking for the right instance - which is just a bit more code - but `frameElement` by definition is an `HTMLElement` ("element" is in the name).

Reply:
Do you still think I should modify it here or do you agree with the type cast here?

Reply:
For example, using `frameElement instanceof window.parent.HTMLElement` can result in:

```
DOMException: Failed to read a named property 'HTMLElement' from 'Window': Blocked a frame with origin "http://localhost:5151" from accessing a cross-origin frame.
```

Reply:
Resolved in https://github.com/bokeh/bokeh/pull/14003/commits/c639a85a7a571d0075e78b6804613d9af6ecea99